bee-node/src/core.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Contains the base types to build a Bee node.
+
+use crate::shutdown::ShutdownTx;
+
+use bee_runtime::{node::Node, worker::Worker};
+
+use anymap::{any::Any as AnyMapAny, Map};
+use futures::Future;
+use fxhash::FxBuildHasher;
+use tokio::task;
+
+use std::{
+    any::{type_name, TypeId},
+    collections::{HashMap, HashSet},
+    pin::Pin,
+};
+
+pub(crate) type WorkerStart<N> = dyn for<'a> FnOnce(&'a mut N) -> Pin<Box<dyn Future<Output = ()> + 'a>>;
+pub(crate) type WorkerStop<N> = dyn for<'a> FnOnce(&'a mut N) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> + Send;
+pub(crate) type ResourceRegister<N> = dyn for<'a> FnOnce(&'a mut N);
+pub(crate) type WorkerStopMap<N> = HashMap<TypeId, Box<WorkerStop<N>>>;
+pub(crate) type WorkerNameMap = HashMap<TypeId, &'static str>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CoreError {
+    #[error("storage backend operation failed: {0}")]
+    StorageBackend(Box<dyn std::error::Error>),
+    #[error("shutdown error")]
+    Shutdown,
+}
+
+/// Encapsulates worker, resource, task management and shutdown logic.
+#[allow(clippy::type_complexity)]
+pub struct Core<N: Node> {
+    pub(crate) workers: Map<dyn AnyMapAny + Send + Sync>,
+    pub(crate) tasks: HashMap<
+        TypeId,
+        Vec<(
+            ShutdownTx,
+            // TODO Result ?
+            Box<dyn Future<Output = Result<(), task::JoinError>> + Send + Sync + Unpin>,
+        )>,
+    >,
+    pub(crate) resources: Map<dyn AnyMapAny + Send + Sync>,
+    pub(crate) worker_stops: HashMap<TypeId, Box<WorkerStop<N>>>,
+    pub(crate) worker_order: Vec<TypeId>,
+    pub(crate) worker_names: HashMap<TypeId, &'static str>,
+}
+
+impl<N: Node> Core<N> {
+    /// Creates a new node core.
+    pub(crate) fn new(worker_stops: WorkerStopMap<N>, worker_order: Vec<TypeId>, worker_names: WorkerNameMap) -> Self {
+        log_topological_order(&worker_order, &worker_names);
+
+        Self {
+            workers: Map::new(),
+            tasks: HashMap::new(),
+            resources: Map::new(),
+            worker_stops,
+            worker_order,
+            worker_names,
+        }
+    }
+
+    /// Adds a worker to the node core.
+    pub(crate) fn add_worker<W: Worker<N> + Send + Sync>(&mut self, worker: W) {
+        self.workers.insert(worker);
```

Comment:
They are, but the architecture of this code can be very confusing. Note, that this is not a fundamental redesign of the bee-node code, but instead an equivalent refactor with the goal to make use of the new bee-autopeering crate, and share as much code as possible between the `fullnode` and the new `entrynode` type. I think this crate can and has to be rewritten from the ground-up so that it becomes much easier to understand, but that's another PR.

Reply:
To answer the question: The purpose of this method is not to add arbitrary workers at runtime. It's supposed to be called only in the way it is done currently, i.e. by a `*NodeBuilder` type, which is passed to all subsystems that need to spawn workers onto the common node core.